### ISSUE TITLE
Character Count for SEO Page Titles and Description

### ISSUE THREAD

**Issue by skramio** [dnnsoftware/Dnn.Platform#3342]
Would be great to have a live character counter at the end of the page title and description fields that corresponds with the recommended max length. The counter could be green when the user has less than the max and red once they've gone over.

**Comment by thabaum** [comment]
I have created javascript that does this. Looking for how to put it into the platform.

![image](https://user-images.githubusercontent.com/13577556/69680565-20b72600-1060-11ea-979c-36c81b40edbd.png)

I have a js script with number that would be next to the help icon that changes from green to red after so many characters.  I just need to know where to put the JS and set the ID's on the form so I can edit the JS to work with those fields only.  I thought I might put one for the keywords as well just so you can see character count on those three site settings.

- next to info tooltip?  this fits the screen probably the best without adding spacing one way or the other up or down.
-  To the right of the textareas
- below the text area with a language resource of # of Characters
-  if there was a maxLegth value i could set this as well to show number of characters remaining but this is bound to change. 

I am just looking up best practices for when to turn the number green or red so you can go over it if you like.

Any support here to help me make a solid pull request that might get accepted is welcome as I have the javascript portion ready to be added for testing further.

**Comment by thabaum** [comment]
I have spent some time today looking at how this all stacks together... what a mind full getting personaBar trying to add JS script to it... There is no place like home, and I am far from there looking at this.

My Current Current Tree Patch Fork

I added a couple language variables for Characters and Characters Remaining for some future work.

I am thinking of TITLE : HELP : COUNTER with the text area below as it is.

So looking at my changes in Javascript and the JSX file (first time there) I am hoping I am able to add "id" to some labels to pass the colored number to.

Is there a fast way I can test this such as just putting these edited files in an installation running?

Pretty much all the logic is there just need to link it all together...

Thank you all for your support!

**Comment by Timo-Breumelhof** [comment]
I would add the counter behind the Title, wrap the counter in a span and add a class, so users can hide this effect if they want, by overruling the CSS.
(like for an intranet where it's not really relevant)

**Comment by thabaum** [comment]
I was going to put it in a span, that was my original idea, I am looking at the code wondering what is ok to do/not to do.

so after label i was originally going to put <span id="siteTitleCounter" class="site-title-counter"></span>

Wouldn't an ID make it select-able in CSS?  I have set an id for the labels created set but the span I can put a class to make it easier to set as well like above.  This makes sense to me as well. These forms are pretty code bound so you really have to dig to find things.

I did put an ID class for selecting the new labels but I am all for adding a CSS class and changing it to a SPAN.   I am just feeling things out and trying to understand what would work the best.

This is nearly my first time working with javascript let alone DNN.  I generally just go for code already creating thinking the best is already done why reinvent the wheel.  But now I see there are so many different types of wheels all for different vehicles.

I don't see any DIV or SPAN in the file so I don't want to do something undesired.

Let me know what you guys want I will do my best to make it happen.  I am just looking at how things are grouped up. 

Also this is just for Site Settings, it is an admin option.  Should not bother anyone too much but if it does I like being able to style/hide it if you like.

**Comment by thabaum** [comment]
@Timo-Breumelhof 
`                    <span
                        id="siteKeyWordsCount"
                        class="site-keywords-count"
                    <Label label={resx.get("Characters.Text") } /> 
                        />`
Is this what you are thinking?

**Comment by Timo-Breumelhof** [comment]
something like that but I would add this to Titles and Description, not the keywords (nobody really uses them any more AFAIK)
I'm not sure I would use a extra label, just a span seems enough (?)

**Comment by thabaum** [comment]
the other label was for the characters but that can be put in the css.

I will look at some examples to see what wrapping the Label would do.  I am not looking to shift anything around.  I was just going to add the span.  And the Characters.Text could be added in the JS file. Or not have any words and just a color changing number.  Which I like that idea as it is the simplest, however if you want to show a user they only have so many words left I need to learn how to do this as well for DNN.  Even if I don't do it in the current PR.  
I can change where they are put in the form like this.  And like you said maybe wrap them both in a span... I am just wondering where the number will show up in regards to the description let me try it.
I was thinking it would be best to put this after the label.
`<span id="siteKeyWordsCount" class="site-keywords-count" />`

to wrap a label I would wish to see the result first so let me test it if I can.

**Comment by thabaum** [comment]
I set it like this for now as I am not sure how it will act wrapping around the Label react component
`                    <Label
                        tooltipMessage={resx.get("plDescription.Help") }
                        label={resx.get("plDescription") }
                    />
                    <span
                        id="siteDescriptionCount"
                        class="site-description-count"
                    />`

And I added an id to the text areas.  And put the JS to handle it in the SiteSettings.js file.  I am sure it can all be done there on the JSX file.  This is done however maybe I could use some comments in the sitesettings.js file so you can skip over that code or create a new file or refactor to React?

**Comment by valadas** [comment]
We do have a reusable component already for that in use for the password fields when creating a new user. It has a red-yellow-green progress bar and a reusable label on the right of the fields. I should weak, fair or strong in the password. We can re-purpose it by using short, acceptable, optimal, too long, or some such, having red when very far from optimal, yellow when near optimal and green when very close to optimal...

**Comment by thabaum** [comment]
@valadas sounds like a winner.  I think maybe a combination of the two so you can see how many characters as well for those trying to make sure things stay within their own set limits if needed.  Maybe we could wrap that label with something similar to what this does?  

I think it would be great to have two more labels like that bar one for character count and another for characters remaining that can be included anywhere in a DNN module.  That would be very cool IMO.

**Comment by skramio** [comment]
Hi all - I was talking with Mitch today about a different item and this topic popped back into my head. Wanted to bump it accordingly.

**Comment by skramio** [comment]
Wanted to reach out and see if there is anything I can do to help possibly get this in place? Looks like a few of you have some great ideas on how to implement - would love to see this rolled out!

**Comment by valadas** [comment]
Well, the issue is pretty well documented, the project being open source, we are just waiting on a code contribution for it.
